src/System.Private.CoreLib/shared/System/BitOps.cs

```diff
@@ -8,31 +8,60 @@
     {
+        // C# optimization that directly wraps the data section of the dll (a bit like string constants)
+        // https://github.com/dotnet/roslyn/pull/24621
+
+        private static ReadOnlySpan<byte> TrailingCountMultiplyDeBruijn => new byte[32]
+        {
+            00, 01, 28, 02, 29, 14, 24, 03,
+            30, 22, 20, 15, 25, 17, 04, 08,
+            31, 27, 13, 23, 21, 19, 16, 07,
+            26, 12, 18, 06, 11, 05, 10, 09
+        };
+
+        /// <summary>
+        /// Count the number of trailing zero bits in an integer value.
+        /// Similar in behavior to the x86 instruction TZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint TrailingZeroCount(int value)
+            => TrailingZeroCount(unchecked((uint)value));
+
+        /// <summary>
+        /// Count the number of trailing zero bits in an integer value.
+        /// Similar in behavior to the x86 instruction TZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static int TrailingZeroCount(int matches)
+        public static uint TrailingZeroCount(uint value)
         {
             if (Bmi1.IsSupported)
             {
-                return (int)Bmi1.TrailingZeroCount((uint)matches);
-            }
-            else // Software fallback
-            {
-                // https://graphics.stanford.edu/~seander/bithacks.html#ZerosOnRightMultLookup
-                // uint.MaxValue >> 27 is always in range [0 - 31] so we use Unsafe.AddByteOffset to avoid bounds check
-                return Unsafe.AddByteOffset(
-                    ref MemoryMarshal.GetReference(TrailingCountMultiplyDeBruijn),
-                    ((uint)((matches & -matches) * 0x077CB531U)) >> 27);
+                // Note that TZCNT contract specifies 0->32
+                return Bmi1.TrailingZeroCount(value);
             }
-        }
 
-        private static ReadOnlySpan<byte> TrailingCountMultiplyDeBruijn => new byte[32]
-        {
-            0, 1, 28, 2, 29, 14, 24, 3, 30, 22, 20, 15, 25, 17, 4, 8,
-            31, 27, 13, 23, 21, 19, 16, 7, 26, 12, 18, 6, 11, 5, 10, 9
-        };
+            const uint deBruijn = 0x077C_B531u;
+            uint ix = (uint)((value & -value) * deBruijn) >> 27;
+
+            // uint.MaxValue >> 27 is always in range [0 - 31] so we use Unsafe.AddByteOffset to avoid bounds check
+            ref byte tz = ref MemoryMarshal.GetReference(TrailingCountMultiplyDeBruijn);
+            uint count = Unsafe.AddByteOffset(ref tz, (IntPtr)ix);
+
+            // Above code has contract 0->0, so we need to special-case 0
+            // Branchless equivalent of: c32 = value == 0 ? 32 : 0
+            bool is0 = value == 0;
+            uint c32 = Unsafe.As<bool, byte>(ref is0) * 32u;
+
+            return c32 + count;
```

Comment:
This is the core fix this PR is proposing. My aim is that this is a branchless equivalent of:
```csharp
if (value == 0)
    return 32u;
```
Obviously the proposed fix is executed post-facto vs the idiomatic guard clause above, but it saves a branch.
Please confirm that my idea is sound - I am not sure how to get the JIT to emit a CMOVNZ but I think this comes close.

Reply:
What is the actual codegen here?

I also wonder if it would be possible to just fixup the debruijn table to account for this....

Reply:
```asm
test edx, edx
setz al
movzx eax, al
shl eax, 0x5
add eax, r8d
ret
```
`return value == 0 ? 32 : count;` is
```asm
test edx, edx
jz L0008
mov eax, r8d
ret
mov eax, 0x20
ret
```
And maybe at some point recognised as a `cmov` https://github.com/dotnet/coreclr/pull/16156

Reply:
> I also wonder if it would be possible to just fixup the debruijn table

I spent last night reading through the deBruijn papers, trying to understand how it works. I think I get it now, it's just a clever perfect hash scheme using 'overlapped' storage. 

Since it's a perfect hash, we need all 32 members for legitimate hashes, so we'd have to add a 33rd member to the array for the special case. 
`Log2(1) = 0`, so we do need the [0] member, we can't store `32` in this slot
`Log2(0) = undefined`, so we need a [32] member

Due to the construction requirements of deBruijn (`2^n`), that means we'd actually need a 64-member array which is wasteful but would in fact work.

(Someone smarter than I may correct me here)

Reply:
> shl eax, 0x5

Ah. I will make that explicit in C#